### PR TITLE
(HC-55) (HC-42) (HC-52) Various HOCON work

### PR DESCRIPTION
**Pull request by HAIL9000**
This pull request addresses several pieces of several HOCON tickets, including:

1. (HC-55) wire config_concatenation into parse_context
2. (HC-42) Add minimum implementation of resolver 
3. (HC-52) Implement config_reference and friends

### CODE REVIEW

File: lib/inc/internal/nodes/config_node_concatenation.hpp [hunk 1]
```diff
@@ -11,4 +11,6 @@ namespace hocon {
         std::shared_ptr<config_node_complex_value> new_node(shared_node_list nodes) override;
     };
 
+    using shared_node_concatenation = std::shared_ptr<config_node_concatenation>;
```
Comment (MikaelSmith) [inline lib/inc/internal/nodes/config_node_concatenation.hpp]: I've started putting these in `types.hpp` with forward declarations to reduce header explosions.

File: lib/src/config.cc [hunk 2]
```diff
@@ -46,9 +47,13 @@ namespace hocon {
     }
 
     shared_config config::resolve_with(shared_config source, config_resolve_options options) const {
-        // TODO: implement all the resolve classes, then this method
-        // until then DO NOT TRY TO PARSE INPUT WITH SUBSTITUTIONS
-        return shared_from_this();
+        auto resolved = resolve_context::resolve(_object, source->_object, move(options));
+
+        if (resolved == _object) {
+            return const_pointer_cast<config>(shared_from_this());
```
Comment (MikaelSmith) [inline lib/src/config.cc]: Why do you have to const_pointer_cast here?

File: lib/src/config.cc [hunk 3]
```diff
@@ -46,9 +47,13 @@ namespace hocon {
...
+        } else {
+            return make_shared<config>(dynamic_pointer_cast<const config_object>(resolved));
```
Comment (MikaelSmith) [inline lib/src/config.cc]: This uses the constructor, rather than `config_object::to_config`?

File: lib/src/config_parser.cc [hunk 4]
```diff
@@ -198,6 +198,21 @@ namespace hocon { namespace config_parser {
         return make_shared<simple_config_object>(object_origin, move(values));
     }
 
+    shared_value parse_context::parse_concatenation(shared_ptr<config_node_concatenation> n) {
+        if (_flavor == config_syntax::JSON) {
+            throw config_exception("Found a concatenation node in JSON");
```
Comment (MikaelSmith) [inline lib/src/config_parser.cc]: We've implemented bug_or_broken_exception, we should try to follow the semantics of the original exception types.

File: lib/inc/internal/nodes/config_node_concatenation.hpp (hunk 1 above)
Comment (branan) [inline lib/inc/internal/nodes/config_node_concatenation.hpp]: `config_node_concatenation` is in `internal` and `types.hpp` is public

File: lib/src/config_parser.cc (hunk 4 above)
Comment (branan) [inline lib/src/config_parser.cc]: Yup, this was some of my early bits and predates those exceptions

File: lib/inc/internal/nodes/config_node_concatenation.hpp (hunk 1 above)
Comment (MikaelSmith) [inline lib/inc/internal/nodes/config_node_concatenation.hpp]: Ah, decent reason.

File: lib/src/resolve_context.cc [hunk 5]
```diff
@@ -1,14 +1,19 @@
...
 #include <internal/resolve_context.hpp>
 #include <internal/resolve_result.hpp>
 #include <internal/resolve_source.hpp>
 
+using namespace std;
+
 namespace hocon {
 
+    resolve_context::resolve_context(config_resolve_options options, path restrict_to_child)
+         : _options(move(options)), _restrict_to_child(restrict_to_child) { }
```
Comment (MikaelSmith) [inline lib/src/resolve_context.cc]: Could be `move`d.

File: lib/src/resolve_source.cc [hunk 6]
```diff
@@ -1,14 +1,102 @@
...
+
+        } else {
+            throw bug_or_broken_exception("resolved object to non-object");
+        }
+
+    }
+
+    resolve_source::value_with_path resolve_source::find_in_object(shared_object obj, path the_path) const {
+        // TODO: implement improve_not_resolved
+        return find_in_object(obj, the_path, {});
```
Comment (MikaelSmith) [inline lib/src/resolve_source.cc]: Should we throw an exception here?

File: lib/src/resolve_source.cc [hunk 7]
```diff
@@ -1,14 +1,102 @@
...
+
+    resolve_source::value_with_path resolve_source::find_in_object(shared_object obj, path the_path, node parents) const {
+        auto key = the_path.first();
+        auto next = the_path.remainder();
+
+        auto v = obj->attempt_peek_with_partial_resolve(*key);
+
+        parents.push_front(dynamic_pointer_cast<const container>(obj));
+
+        if(next.empty()) {
```
Comment (MikaelSmith) [inline lib/src/resolve_source.cc]: Linter probably complained about the missing space here.

File: lib/src/values/simple_config_object.cc [hunk 8]
```diff
@@ -7,23 +7,43 @@
...
+#include <algorithm>
+
 using namespace std;
 
 namespace hocon {
 
     struct simple_config_object::resolve_modifier : public modifier {
-        // TODO: initialize origin_restrict
-        resolve_modifier(resolve_context c, resolve_source s) : context(move(c)), source(move(s)), origin_restrict("") {}
+        // TODO: initialize original_restrict
```
Comment (MikaelSmith) [inline lib/src/values/simple_config_object.cc]: Looks like this comment could go away.

File: lib/src/values/simple_config_object.cc [hunk 9]
```diff
@@ -32,6 +52,16 @@ namespace hocon {
         config_object(move(origin)), _value(move(value)), _resolved(status), _ignores_fallbacks(ignores_fallbacks)
     {}
 
+    simple_config_object::simple_config_object(shared_origin origin,
+                                               unordered_map <std::string, shared_value> value)
+         : config_object(move(origin)) {
+        // These are in the body so I can call resolve_from_status
```
Comment (MikaelSmith) [inline lib/src/values/simple_config_object.cc]: Not sure this is needed to ensure a well-defined order.

File: lib/tests/CMakeLists.txt [hunk 10]
```diff
@@ -10,7 +11,7 @@ set(TEST_CASES
     tokenizer_test.cc
     test_utils.cc
     config_document_tests.cc
-    conf_parser_test.cc)
+    conf_parser_test.cc config_substitution_test.cc)
```
Comment (MikaelSmith) [inline lib/tests/CMakeLists.txt]: Bleh. Check if you have the latest CLion, I thought mine was handling this better.

**Comment by MikaelSmith** [comment]
Done reviewing. Mostly good, just need to satisfy the linter.

File: lib/src/config.cc (hunk 2 above)
Comment (HAIL9000) [inline lib/src/config.cc]: @branan I'm not sure if I can answer this question... could you help?

**Comment by HAIL9000** [comment]
@MikaelSmith Updated! Fixups were all amened to the last commit

File: lib/src/resolve_source.cc [hunk 11]
```diff
@@ -1,14 +1,97 @@
...
+            return {make_resolve_result(new_context, pair.value), pair.path_from_root};
+
+        } else {
+            throw bug_or_broken_exception("resolved object to non-object");
+        }
+    }
+
+    resolve_source::value_with_path resolve_source::find_in_object(shared_object obj, path the_path) const {
+        // TODO: implement improve_not_resolved
+        throw config_exception("resolve_source::value_with_path has not been implemented");
```
Comment (MikaelSmith) [inline lib/src/resolve_source.cc]: Oh, this exception causes some tests to fail.

**Pull request merged**